pkg/rlptools/wasm_utils.go

```diff
@@ -10,98 +10,100 @@ import (
 	istioclientgoextensionv1alpha1 "istio.io/client-go/pkg/apis/extensions/v1alpha1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	gatewayapiv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	gatewayapiv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
 
-	kuadrantv1beta1 "github.com/kuadrant/kuadrant-operator/api/v1beta1"
+	kuadrantv1beta2 "github.com/kuadrant/kuadrant-operator/api/v1beta2"
 	"github.com/kuadrant/kuadrant-operator/pkg/common"
 )
 
 var (
 	WASMFilterImageURL = common.FetchEnv("RELATED_IMAGE_WASMSHIM", "oci://quay.io/kuadrant/wasm-shim:latest")
+
+	PathMatchTypeMap = map[gatewayapiv1alpha2.PathMatchType]PatternOperator{
+		gatewayapiv1beta1.PathMatchExact:             PatternOperator(kuadrantv1beta2.EqualOperator),
+		gatewayapiv1beta1.PathMatchPathPrefix:        PatternOperator(kuadrantv1beta2.StartsWithOperator),
+		gatewayapiv1beta1.PathMatchRegularExpression: PatternOperator(kuadrantv1beta2.MatchesOperator),
+	}
 )
 
-type GatewayAction struct {
-	Configurations []kuadrantv1beta1.Configuration `json:"configurations"`
+type SelectorSpec struct {
+	// Selector of an attribute from the contextual properties provided by Envoy
+	// during request and connection processing
+	// https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/advanced/attributes
+	// They are named by a dot-separated path (e.g. request.path)
+	// Examples:
+	// "request.path" -> The path portion of the URL
+	Selector string `json:"selector"`
+
+	// If not set it defaults to `selector` field value as the descriptor key.
+	// +optional
+	Key *string `json:"key,omitempty"`
 
+	// An optional value to use if the selector is not found in the context.
+	// If not set and the selector is not found in the context, then no descriptor is generated.
 	// +optional
-	Rules []kuadrantv1beta1.Rule `json:"rules,omitempty"`
+	Default *string `json:"default,omitempty"`
 }
 
-func DefaultGatewayConfiguration(key client.ObjectKey) []kuadrantv1beta1.Configuration {
-	return []kuadrantv1beta1.Configuration{
-		{
-			Actions: []kuadrantv1beta1.ActionSpecifier{
-				{
-					GenericKey: &kuadrantv1beta1.GenericKeySpec{
-						DescriptorValue: key.String(),
-						// using default value as specified in Envoy spec
-						// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-ratelimit-action-generickey
-						DescriptorKey: &[]string{"ratelimitpolicy"}[0],
-					},
-				},
-			},
-		},
-	}
+type StaticSpec struct {
+	Value string `json:"value"`
+	Key   string `json:"key"`
 }
 
-// GatewayActionsFromRateLimitPolicy return flatten list from GatewayAction from the RLP
-func GatewayActionsFromRateLimitPolicy(rlp *kuadrantv1beta1.RateLimitPolicy, route *gatewayapiv1alpha2.HTTPRoute) []GatewayAction {
-	flattenActions := make([]GatewayAction, 0)
-	if rlp == nil {
-		return flattenActions
-	}
+// TODO implement one of constraint
```

Comment:
I see at least a couple ways to implement this. E.g. https://github.com/Kuadrant/authorino/blob/5db6fb9153d5229188b03292249cbf04bfb5fafb/install/crd/kustomization.yaml#L15-L21.

Is that what you had in mind? Or perhaps a validating webhook instead? Or yet, "late" in the process, in the controller, during reconciliation?

Reply:
This is the OpenAPIv3 validation that prevents objects to be created and/or updated without needing any validation webhook.